internal/util/kubernetes/object/set.go

```diff
@@ -1,55 +1,89 @@
+	generation int64
+	succeeded  bool
+	// TODO: add reason/message here when configuration failed.
 }
 
-func (s *Set) Insert(objs ...client.Object) {
-	for _, obj := range objs {
-		if s.store == nil {
-			s.store = make(map[gvk]map[namespace]map[name]struct{})
-		}
+type ConfigurationStatus string
 
-		objGVK := obj.GetObjectKind().GroupVersionKind().String()
-		objNS := obj.GetNamespace()
-		objName := obj.GetName()
+const (
+	ConfigurationStatusSucceeded ConfigurationStatus = "Succeeded"
+	ConfigurationStatusFailed    ConfigurationStatus = "Failed"
+	ConfigurationStatusUnknown   ConfigurationStatus = "Unknown"
+)
 
-		if s.store[gvk(objGVK)] == nil {
-			s.store[gvk(objGVK)] = make(map[namespace]map[name]struct{})
-		}
+// ConfigurationStatusSet is a de-duplicate set to store the configure status
+// (succeeded, failed, unknown) of kubernetes objects.
+type ConfigurationStatusSet struct {
+	store map[gvk]map[types.NamespacedName]objectConfigurationStatus
+}
 
-		if s.store[gvk(objGVK)][namespace(objNS)] == nil {
-			s.store[gvk(objGVK)][namespace(objNS)] = make(map[name]struct{})
-		}
+func NewConfigurationStatusSet() *ConfigurationStatusSet {
+	return &ConfigurationStatusSet{
+		store: map[gvk]map[types.NamespacedName]objectConfigurationStatus{},
+	}
+}
 
-		s.store[gvk(objGVK)][namespace(objNS)][name(objName)] = struct{}{}
+func (s *ConfigurationStatusSet) Insert(obj client.Object, succeeded bool) {
+	if s.store == nil {
+		s.store = make(map[gvk]map[types.NamespacedName]objectConfigurationStatus)
+	}
+
+	objGVK := gvk(obj.GetObjectKind().GroupVersionKind().String())
+	nsName := types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	}
+	if s.store[objGVK] == nil {
+		s.store[objGVK] = make(map[types.NamespacedName]objectConfigurationStatus)
+	}
+	s.store[objGVK][nsName] = objectConfigurationStatus{
+		generation: obj.GetGeneration(),
+		succeeded:  succeeded,
 	}
 }
 
-func (s *Set) Has(obj client.Object) bool {
-	gvkStr := obj.GetObjectKind().GroupVersionKind().String()
-	gvkMap, ok := s.store[gvk(gvkStr)]
+func (s *ConfigurationStatusSet) Get(obj client.Object) ConfigurationStatus {
+	if s.store == nil {
+		return ConfigurationStatusUnknown
+	}
+
+	objGVK := gvk(obj.GetObjectKind().GroupVersionKind().String())
+	nsName := types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	}
+
+	gvkMap, ok := s.store[objGVK]
 	if !ok {
-		return false
+		return ConfigurationStatusUnknown
 	}
 
-	namespaceStr := obj.GetNamespace()
-	namespaceMap, ok := gvkMap[namespace(namespaceStr)]
+	status, ok := gvkMap[nsName]
 	if !ok {
-		return false
+		return ConfigurationStatusUnknown
+	}
+
+	// if the object generation is newer than the generation of current configuration,
+	// the latest specification of the object may still not configured on Kong gateway, so "Unknown" is returned.
+	if status.generation < obj.GetGeneration() {
+		return ConfigurationStatusUnknown
+	}
+
+	if !status.succeeded {
```

Comment:
If the status is unknown at this point, we return `ConfigurationStatusFailed`. Is this an intended behavior?

Reply:
What do you mean by "unknown at this point"? If the object is never inserted, the 
```
	status, ok := gvkMap[nsName]
	if !ok {
		return ConfigurationStatusUnknown
	}
```
before here will return `ConfigurationStatusUnknown`.